storage/pebble/procedure/cluster.go

```diff
@@ -0,0 +1,228 @@
+		}
+
+		err = operation.RetrieveHeader(finalID, final)(tx)
+		if err != nil {
+			return fmt.Errorf("could not retrieve finalized header: %w", err)
+		}
+
+		return nil
+	}
+}
+
+// FinalizeClusterBlock finalizes a block in cluster consensus.
+func FinalizeClusterBlock(blockID flow.Identifier) func(storage.PebbleReaderBatchWriter) error {
+	return func(tx storage.PebbleReaderBatchWriter) error {
+		r, w := tx.ReaderWriter()
+
+		// retrieve the header to check the parent
+		var header flow.Header
+		err := operation.RetrieveHeader(blockID, &header)(r)
+		if err != nil {
+			return fmt.Errorf("could not retrieve header: %w", err)
+		}
+
+		// get the chain ID, which determines which cluster state to query
+		chainID := header.ChainID
+
+		// retrieve the current finalized state boundary
+		var boundary uint64
+		err = operation.RetrieveClusterFinalizedHeight(chainID, &boundary)(r)
+		if err != nil {
+			return fmt.Errorf("could not retrieve boundary: %w", err)
+		}
+
+		// retrieve the ID of the boundary head
+		var headID flow.Identifier
+		err = operation.LookupClusterBlockHeight(chainID, boundary, &headID)(r)
+		if err != nil {
+			return fmt.Errorf("could not retrieve head: %w", err)
+		}
+
+		// check that the head ID is the parent of the block we finalize
+		if header.ParentID != headID {
+			return fmt.Errorf("can't finalize non-child of chain head")
+		}
+
+		// insert block view -> ID mapping
+		err = operation.IndexClusterBlockHeight(chainID, header.Height, header.ID())(w)
+		if err != nil {
+			return fmt.Errorf("could not insert view->ID mapping: %w", err)
+		}
+
+		// update the finalized boundary
+		err = operation.UpdateClusterFinalizedHeight(chainID, header.Height)(w)
+		if err != nil {
+			return fmt.Errorf("could not update finalized boundary: %w", err)
+		}
+
+		// NOTE: we don't want to prune forks that have become invalid here, so
+		// that we can keep validating entities and generating slashing
+		// challenges for some time - the pruning should happen some place else
+		// after a certain delay of blocks
+
+		return nil
+	}
+}
+
+// InsertClusterPayload inserts the payload for a cluster block. It inserts
+// both the collection and all constituent transactions, allowing duplicates.
+func InsertClusterPayload(blockID flow.Identifier, payload *cluster.Payload) func(pebble.Writer) error {
+	return func(tx pebble.Writer) error {
+
+		// cluster payloads only contain a single collection, allow duplicates,
+		// because it is valid for two competing forks to have the same payload.
+		light := payload.Collection.Light()
+		err := operation.InsertCollection(&light)(tx)
+		if err != nil {
+			return fmt.Errorf("could not insert payload collection: %w", err)
+		}
+
+		// insert constituent transactions
+		for _, colTx := range payload.Collection.Transactions {
+			err = operation.InsertTransaction(colTx.ID(), colTx)(tx)
+			if err != nil {
+				return fmt.Errorf("could not insert payload transaction: %w", err)
+			}
+		}
```

Comment:
note that `payload.Collection.Light()` already computes the IDs of the transactions. This has non-negligible cost due to serialization and potentially hashing somewhat larger data structures. 

Lets not needlessly repeat that here. The `colTx.ID()` in line 172 is already pre-computed in `light`